plugins/log4j2-log-plugin/src/main/java/com/megaease/easeagent/log4j2/log/Log4jLogMapper.java

```diff
@@ -0,0 +1,121 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.megaease.easeagent.log4j2.log;
+
+import com.megaease.easeagent.plugin.api.config.IPluginConfig;
+import com.megaease.easeagent.plugin.api.otlp.common.AgentLogData;
+import com.megaease.easeagent.plugin.api.otlp.common.AgentLogDataImpl;
+import com.megaease.easeagent.plugin.api.otlp.common.LogMapper;
+import com.megaease.easeagent.plugin.interceptor.MethodInfo;
+import com.megaease.easeagent.plugin.utils.SystemClock;
+import io.opentelemetry.sdk.logs.data.Severity;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.ThreadContext;
+import org.apache.logging.log4j.message.MapMessage;
+import org.apache.logging.log4j.message.Message;
+
+import java.util.Map;
+
+/**
+ * reference to Opentelemetry instrumentation
+ */
+public class Log4jLogMapper implements LogMapper {
+    private static final String SPECIAL_MAP_MESSAGE_ATTRIBUTE = "message";
+
+    public Log4jLogMapper() {}
+
+    public AgentLogData mapLoggingEvent(MethodInfo logInfo, int levelInt, IPluginConfig config) {
+        // level
+        Level level = (Level)logInfo.getArgs()[0];
+        if (level.intLevel() < levelInt) {
+            return null;
+        }
+
+        AgentLogDataImpl.Builder builder = AgentLogDataImpl.builder();
+        builder.severity(levelToSeverity(level));
+        builder.severityText(level.name());
+
+        // logger
+        Logger logger = (Logger)logInfo.getInvoker();
+        if (logger.getName() == null || logger.getName().isEmpty()) {
+            builder.logger("ROOT");
+        } else {
+            builder.logger(logger.getName());
+        }
+
+        // message
+        Message message = (Message)logInfo.getArgs()[4];
+        if (!(message instanceof MapMessage)) {
+            builder.body(message.getFormattedMessage());
+        } else {
+            MapMessage<?, ?> mapMessage = (MapMessage<?, ?>) message;
+
+            String body = mapMessage.getFormat();
+            boolean checkSpecialMapMessageAttribute = (body == null || body.isEmpty());
+            if (checkSpecialMapMessageAttribute) {
+                body = mapMessage.get(SPECIAL_MAP_MESSAGE_ATTRIBUTE);
+            }
+            if (body != null && !body.isEmpty()) {
+                builder.body(body);
+            }
+        }
+
+
+        // throwable
+        Throwable throwable = (Throwable) logInfo.getArgs()[5];
+        if (throwable != null) {
+            builder.throwable(throwable);
+        }
+
+        // thread
+        builder.thread(Thread.currentThread());
+        builder.epochMills(SystemClock.now());
+
+        // MDC
+        Map<String, String> contextData = ThreadContext.getImmutableContext();
+        builder.contextData(config.getStringList(LogMapper.MDC_KEYS), contextData);
+
+        // span context
+        builder.spanContext();
```

Comment:
logInfo().getArgs()[?] could  throw out of index exception, I don't think the code isn't robust, I recommend that: 
```
Args args = logInfo.getArgs()
for (int i = 0 ;i< args.length; ++i) {
   switch i {
     case 0:
       build.set(args[i]....)
       break;
     case 1:  
       build.set(args[i]....)
       break;
     case 2:
       build.set(args[i]....)
       break;
....
   }
return build.build();
}
```

Reply:
The enhanced method always has 5 parameters, it never causes out of index exception.

Reply:
Even though it always has 5 arguments,  no one can ensure the argument will not be changed in the future. This is is the defensive programming style and it does its best to try to promote your code robustness.

Reply:
OK, I will improve it.